Gravatar/MruImageCache.cs

```diff
@@ -0,0 +1,171 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Drawing;
+using System.Linq;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+
+namespace Gravatar
+{
+    /// <summary>
+    /// Caches most-recently-used images.
+    /// </summary>
+    /// <remarks>
+    /// Decorates an inner cache, delegating to it as needed.
+    /// <para />
+    /// If an image is available in memory, the inner cache can be bypassed.
+    /// </remarks>
+    public sealed class MruImageCache : IImageCache
+    {
+        private readonly int _cleanAtSize;
+        private readonly int _cleanToSize;
+
+        private readonly ConcurrentDictionary<string, Entry> _entryByFileName = new ConcurrentDictionary<string, Entry>();
+        private readonly IImageCache _inner;
+
+        public MruImageCache([NotNull] IImageCache inner, int cleanAtSize = 30, int cleanToSize = 25)
+        {
+            if (cleanToSize >= cleanAtSize)
+            {
+                throw new ArgumentException($"{nameof(cleanAtSize)} must be less than {nameof(cleanToSize)}.");
+            }
+
+            _inner = inner ?? throw new ArgumentNullException(nameof(inner));
```

Comment:
❓ I would expect this wrapper to attach a handler for `_inner.Invalidated`. Is there a reason this is omitted?

Reply:
Will review, thanks.

Reply:
I believe this works as expected. Check where `Invalidated` is defined on this class. It forwards add/remove calls, and delegates responsibility for invalidation to the inner cache.